vms/platformvm/txs/executor/state_changes.go

```diff
@@ -257,10 +251,97 @@ func advanceTimeTo(
 		changes.DeleteExpiry(expiry)
 	}
 
+	if !backend.Config.UpgradeConfig.IsEtnaActivated(newChainTime) {
+		changes.SetTimestamp(newChainTime)
+		return changes, changed, nil
+	}
+
+	// Calculate number of seconds the time is advancing
+	previousChainTime := changes.GetTimestamp()
+	seconds := uint64(newChainTime.Sub(previousChainTime) / time.Second)
+
+	advanceDynamicFeeState(backend.Config.DynamicFeeConfig, changes, seconds)
+	sovsChanged, err := advanceValidatorFeeState(
+		backend.Config.ValidatorFeeConfig,
+		parentState,
+		changes,
+		seconds,
+	)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to advance validator fee state: %w", err)
+	}
+	changed = changed || sovsChanged
+
 	changes.SetTimestamp(newChainTime)
 	return changes, changed, nil
 }
 
+func advanceDynamicFeeState(
+	config gas.Config,
+	changes state.Diff,
+	seconds uint64,
+) {
+	dynamicFeeState := changes.GetFeeState()
+	dynamicFeeState = dynamicFeeState.AdvanceTime(
+		config.MaxCapacity,
+		config.MaxPerSecond,
+		config.TargetPerSecond,
+		seconds,
+	)
+	changes.SetFeeState(dynamicFeeState)
+}
+
+// advanceValidatorFeeState advances the validator fee state by [seconds]. SoVs
+// are read from [parentState] and written to [changes] to avoid modifying state
+// while an iterator is held.
+func advanceValidatorFeeState(
+	config fee.Config,
+	parentState state.Chain,
+	changes state.Diff,
+	seconds uint64,
+) (bool, error) {
+	validatorFeeState := fee.State{
+		Current: gas.Gas(changes.NumActiveSubnetOnlyValidators()),
+		Excess:  changes.GetSoVExcess(),
+	}
+	validatorCost := validatorFeeState.CostOf(config, seconds)
+
+	accruedFees := changes.GetAccruedFees()
+	accruedFees, err := math.Add(accruedFees, validatorCost)
+	if err != nil {
+		return false, fmt.Errorf("could not calculate accrued fees: %w", err)
+	}
+
+	// Invariant: It is not safe to modify the state while iterating over it,
+	// so we use the parentState's iterator rather than the changes iterator.
+	// ParentState must not be modified before this iterator is released.
+	sovIterator, err := parentState.GetActiveSubnetOnlyValidatorsIterator()
+	if err != nil {
+		return false, fmt.Errorf("could not iterate over active SoVs: %w", err)
+	}
+	defer sovIterator.Release()
+
+	var changed bool
+	for sovIterator.Next() {
+		sov := sovIterator.Value()
+		if sov.EndAccumulatedFee > accruedFees {
```

Comment:
I'm not 100% following how the math works out such that `EndAccumulatedFee` being:
```
equal to the amount of fees this validator is willing to pay plus the amount of globally accumulated fees when this validator started validating
```
also makes it such that
```
EndAccumulatedFee is the amount of globally accumulated fees that can accrue before this validator must be deactivated.
```

Thinking out loud...if the current global accumulated fee is 1 AVAX, and I register a validator with a balance of 1 AVAX, then my `EndAccumulatedFee` should be 2 AVAX. If advancing time 1 minute would require a cumulative of more than 1 AVAX to be charged across all validators, how is it that we know my validator would need to be charged more than 1 AVAX?

Reply:
The wording around `globally accumulated` is confusing. `accruedFees` is the fees that would have been paid by a validator that has been active since the Etna activation. _Not_ the total amount of fees that have been charged to all validators for all time.